gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsTracer.java

```diff
@@ -42,7 +51,111 @@
 @InternalApi
 public class MetricsTracer implements ApiTracer {
 
-  public MetricsTracer(MethodName methodName, MetricsRecorder metricsRecorder) {}
+  public static final String STATUS_ATTRIBUTE = "status";
+
+  private Stopwatch attemptTimer;
+
+  private final Stopwatch operationTimer = Stopwatch.createStarted();
+
+  private final Map<String, String> attributes = new HashMap<>();
+
+  protected MetricsRecorder metricsRecorder;
+
+  public MetricsTracer(MethodName methodName, MetricsRecorder metricsRecorder) {
+    this.attributes.put("method_name", methodName.toString());
+    this.metricsRecorder = metricsRecorder;
+  }
+
+  @Override
+  public void operationSucceeded() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.OK.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationCancelled() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.CANCELLED.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    attributes.put(STATUS_ATTRIBUTE, extractStatus(error));
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    // no-op
+  }
+
+  @Override
+  public void attemptStarted(Object request, int attemptNumber) {
+    attemptTimer = Stopwatch.createStarted();
+  }
+
+  @Override
+  public void attemptSucceeded() {
+
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.OK.toString());
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptCancelled() {
+
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.CANCELLED.toString());
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
```

Comment:
What is this argument `Duration delay`?

Reply:
[This](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/tracing/ApiTracer.java#L115) is the amount of time to wait before the next attempt will start.